rate/limit.go

```diff
@@ -0,0 +1,72 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package rate
+
+import (
+	"context"
+
+	lru "github.com/hashicorp/golang-lru"
+	r "golang.org/x/time/rate"
+)
+
+const (
+	cacheSize       = 1000
+	burstMultiplier = 2
+)
+
+type LimitHandler interface {
```

Comment:
could we call it a `LimitManager` or similar instead of `LimitHandler`? `handler` gets confusing with all the `http.Handler`s we have.